batched/system.go

```diff
@@ -0,0 +1,335 @@
+package batched
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/ipfs/go-ipfs-provider/queue"
+	"strconv"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	provider "github.com/ipfs/go-ipfs-provider"
+	"github.com/ipfs/go-ipfs-provider/simple"
+	logging "github.com/ipfs/go-log"
+	"github.com/ipfs/go-verifcid"
+	"github.com/multiformats/go-multihash"
+)
+
+var log = logging.Logger("provider.batched")
+
+type BatchProvidingSystem struct {
+	ctx   context.Context
+	close context.CancelFunc
+
+	reprovideInterval time.Duration
+	rsys              provideMany
+	keyProvider       simple.KeyChanFunc
+
+	q  *queue.Queue
+	ds datastore.Batching
+
+	provch, dynamicCh chan cid.Cid
+
+	totalProvides, lastReprovideBatchSize     int
+	avgProvideDuration, lastReprovideDuration time.Duration
+}
+
+var _ provider.System = (*BatchProvidingSystem)(nil)
+
+type provideMany interface {
+	ProvideMany(ctx context.Context, keys []multihash.Multihash) error
+	Ready() bool
+}
+
+// Option defines the functional option type that can be used to configure
+// BatchProvidingSystem instances
+type Option func(system *BatchProvidingSystem) error
+
+var lastReprovideKey = datastore.NewKey("/provider/reprovide/lastreprovide")
+
+func New(provider provideMany, q *queue.Queue, opts ...Option) (*BatchProvidingSystem, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &BatchProvidingSystem{
+		ctx:   ctx,
+		close: cancel,
+
+		reprovideInterval: time.Hour * 24,
+		rsys:              provider,
+		keyProvider:       nil,
+		q:                 q,
+		ds:                datastore.NewMapDatastore(),
+		provch:            make(chan cid.Cid, 1),
+		dynamicCh:         make(chan cid.Cid, 1),
+	}
+
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+
+	return s, nil
+}
+
+func Datastore(batching datastore.Batching) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.ds = batching
+		return nil
+	}
+}
+
+func ReproviderInterval(duration time.Duration) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.reprovideInterval = duration
+		return nil
+	}
+}
+
+func KeyProvider(fn simple.KeyChanFunc) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.keyProvider = fn
+		return nil
+	}
+}
+
+func (s *BatchProvidingSystem) Run() {
+	go func() {
+		m := make(map[cid.Cid]struct{})
+		for {
+			pauseDetectTimer := time.NewTimer(time.Hour)
+			maxDurationCollectionTimer := time.NewTimer(time.Minute * 10)
+
+			performedReprovide := false
+		loop:
+			for {
+				select {
+				case c := <-s.provch:
+					m[c] = struct{}{}
+					pauseDetectTimer.Reset(time.Millisecond * 500)
```

Comment:
Is there some code you've used for this before that I can take a look at? We can wait on the issue, but it is a little concerning that if I'm reading 1M CIDs off of disk to provide that I could be resetting the timer 1M times pretty quickly here (i.e. since I'll read all the CIDs in under the 10 minute `maxCollectionDuration`.

Reply:
Not quite the same, but close: https://github.com/libp2p/go-yamux/blob/391ef8d58091a341017faad0172abad689041588/session.go#L464-L472

Basically, I check to see if we've elapsed half the time and only extend the timer if that's the case. But it _probably_ isn't an issue (and we can test).

Reply:
SGTM, we'll defer this change until later if it becomes necessary